### PR TITLE
[dynamo] Remove `mutable_local.source` and index on `VariableTracker` rather than `MutableLocalBase`

### PR DESCRIPTION
**Pull request by StrongerXi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137906
* __->__ #137905

This patch addresses parts of the side-effect refactor proposed in #133027;
specifically, it does 3 things:

1. Change `SideEffects.store_attr_mutations` and `PyCodegen.tempvars`
   to index on `VariableTracker` rather than `MutableLocalBase`.
2. Remove the `source` field from `MutableSideEffects` and
   `AttributeMutation`, and use `VariableTracker.source` instead.
3. Plumb a `overridden_sources: Dict[Source, Source]` from
   `handle_aliases_for_stolen_lists` to `PyCodegen` so that we don't
   update `VariableTracker.source` in place, while still preserving what
   `handle_aliases_for_stolen_lists` needed (i.e., modifying codegen for
   certain `VariableTracker`).

(1) and (2) are merged in 1 patch because of some dependency between
a. `OutputGraph.handle_aliases_for_stolen_lists` which iterates over
   `sideSideEffects.store_attr_mutations.keys()`, and potentially update
   its source field to be completely different.
b. `SideEffects.codegen_update_mutated`, which happens after the above
   and uses `cg(var.mutable_local.source)`.
where if we apply (1) only, (b) breaks, and if we apply (2) only, (a)
breaks.

(3) is needed for correctness, see comments in the PR for details.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

**Comment by StrongerXi** [comment]
Looking into test failures, reproducing has been a bit difficult.

File: torch/_dynamo/side_effects.py [hunk 1]
```diff
@@ -403,6 +397,10 @@ def is_live(var: Union[MutableLocalBase, VariableTracker]):
         # during a graph break (tx.symbolic_locals), and mutation on pre-existing variables.
         # Recursively visit Variables and see if any of them have been mutated.
         VariableTracker.visit(visit, (tx.stack, tx.symbolic_locals, pre_existing_vars))
+        # Manually release the self-referential function, which indirectly
+        # captures certain `VariableTracker` and affects parts of PT test/logic
+        # that are sensitive to when certain objects get released.
+        del visit
```
Comment (StrongerXi) [inline torch/_dynamo/side_effects.py:403]: https://github.com/pytorch/pytorch/pull/136891#discussion_r1781370708 strikes again.

I looked more into our code, and noticed a few similar things regarding manual memory management:
1. https://github.com/pytorch/pytorch/blob/e9d420c8f694e085b2d86e2300f2d3a7487cad09/torch/_dynamo/symbolic_convert.py#L1105-L1109
2. #100282

Looks like there might be more we could push along the above 2, but this "local self-referential function capturing a local set of `VariableTracker`" feels more difficult to solve cleanly. `del visit` is the shortest solution I can think of for now, and I feel like all other local solutions require some chunk of comment like the above to explain why we need such special handling.

I thought about solving it on the `VariableTracker` side, like holding a weakref to the tensor, but then that might end up releasing the tensors too early (right after entering Dynamo analysis)? @jansel do you have any opinion?

**Comment by StrongerXi** [comment]
Fix lint issues and a read after write across `VariableTracker.source` issue in `handle_aliases_for_stolen_lists` now that `source` is associated with a `VariableTracker`.

**Comment by StrongerXi** [comment]
Trying to merge sources exposed some subtleties in `handle_aliases_for_stolen_lists`, will write up a summary.

**Review by jansel — CHANGES_REQUESTED** [review]
Looks like tests are failing.  Also, thinking a bit more, this may be needed for correctness.  There are potentially two sources for something, one before the mutation one after the mutation.

**Comment by StrongerXi** [comment]
> Looks like tests are failing.

I think I got to the bottom of this.

The [test failure](https://hud.pytorch.org/pr/pytorch/pytorch/137905#31523580209) looks like this:

```
  File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/_dynamo/convert_frame.py", line 825, in _compile_inner
    check_fn = CheckFunctionManager(
               ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/_dynamo/guards.py", line 2295, in __init__
    guard.create(builder)
  File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/_guards.py", line 298, in create
    return self.create_fn(builder, self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/_dynamo/guards.py", line 1720, in DUPLICATE_INPUT
    self.get_guard_manager_from_source(source_b),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/_dynamo/guards.py", line 846, in get_guard_manager_from_source
    example_value = self.get(source_name)
                    ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/_dynamo/guards.py", line 1179, in get
    return eval(name, self.scope, _get_closure_vars())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 1, in <module>
torch._dynamo.exc.InternalTorchDynamoError: KeyError: inputs_ref_1
```

Here's what I found.

First add this patch for some debug printing:
```diff
diff --git a/torch/_dynamo/variables/builder.py b/torch/_dynamo/variables/builder.py
index 4d73d3d4fce..f16bfb5bd14 100644
--- a/torch/_dynamo/variables/builder.py
+++ b/torch/_dynamo/variables/builder.py
@@ -368,6 +368,8 @@ class VariableBuilder:
             side_effect_result = self.tx.output.side_effects[value]
             dup_guard = make_dupe_guard(self.source, side_effect_result.source)
             if dup_guard:
+                if str(self.source) == "GetItemSource(base=UnspecializedBuiltinNNModuleSource(base=UnspecializedParamBufferSource(base=AttrSource(base=GetItemSource(base=LocalSource(local_name='hooks', cell_or_freevar=False), index=0, index_is_slice=False), member='mod0'), member='_parameters')), index='weight', index_is_slice=False)":
+                    breakpoint()
                 self.install_guards(dup_guard)
             return side_effect_result
```

Then run the following with & without this patch.

```bash
$ pytest test/inductor/test_distributed_patterns.py::DistributedPatternTests::test_fake_distributed_aot_eager
```

Shared context with & without this patch:
```
[52]   /Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/symbolic_convert.py(3064)_return()
-> self.output.compile_subgraph(
[53]   /Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/output_graph.py(1128)compile_subgraph()
-> self.codegen_suffix(tx, stack_values, pass1)
[54]   /Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/output_graph.py(1187)codegen_suffix()
-> self.side_effects.codegen_update_mutated(cg)
[55]   /Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/side_effects.py(607)codegen_update_mutated()
-> cg(var, allow_cache=False)
[56]   /Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/codegen.py(231)__call__()
-> self.call_reconstruct(value)
[57]   /Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/codegen.py(84)call_reconstruct()
-> res = value.reconstruct(self)
[58]   /Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/variables/dicts.py(225)reconstruct()
-> self._maybe_realize(self.original_items.get(key.vt)) != value.realize()
[59]   /Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/variables/dicts.py(207)_maybe_realize()
-> return item.realize() if item else item
[60]   /Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/variables/lazy.py(63)realize()
-> self._cache.realize()
[61]   /Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/variables/lazy.py(29)realize()
-> self.vt = VariableBuilder(tx, self.source)(self.value)
[62] > /Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/variables/builder.py(373)__call__()
-> self.install_guards(dup_guard)
(Pdb++) l
368                 side_effect_result = self.tx.output.side_effects[value]
369                 dup_guard = make_dupe_guard(self.source, side_effect_result.source)
370                 if dup_guard:
371                     if str(self.source) == "GetItemSource(base=UnspecializedBuiltinNNModuleSource(base=UnspecializedParamBufferSource(base=AttrSource(base=GetItemSource(base=LocalSource(local_name='hooks', cell_or_freevar=False), index=0, index_is_slice=False), member='mod0'), member='_parameters')), index='weight', index_is_slice=False)":
372                         breakpoint()
373  ->                 self.install_guards(dup_guard)
374                 return side_effect_result
375
376             cached_vt = self.tx.output.variable_tracker_cache.lookup(value, self.source)
377             if cached_vt:
378                 return cached_vt
```

Without this patch:
```
(Pdb++) p side_effect_result.source
GetItemSource(base=LocalSource(local_name='inputs', cell_or_freevar=False), index=3, index_is_slice=False)
(Pdb++) p side_effect_result.mutable_local.source
LocalSource(local_name='inputs_ref_1', cell_or_freevar=False)
```

With this patch (note `side_effect_result.mutable_local` no longer has a source):
```
(Pdb++) p side_effect_result.source
LocalSource(local_name='inputs_ref_1', cell_or_freevar=False)
```

In summary:
1. `handle_aliases_for_stolen_lists` mutated a `MutableLocal.source` field, which is needed for proper codegen (more details in #123359)
https://github.com/pytorch/pytorch/blob/0b7ef196cd80267285c4c5e0117ff1df50cb5710/torch/_dynamo/output_graph.py#L973
2. this wasn't an issue because `VariableTracker.source` was still the original source (see the pdb trace above), and that got used correctly through all the guards logic.
3. this patch merges `MutableLocal.source `into `VariableTracker.source`, which means we no longer have the original source anywhere, which caused a problem in the guards logic, i.e., can't find `inputs_ref_1` in the python scope.

To me (2) feels a bit subtle (or maybe there are invariants/rules about using `Source` I'm not aware of).

**Comment by StrongerXi** [comment]
I have some ideas for a temporary workaround, but I also wanted to see if we could find a better solution for #123359, as part of the side-effect modeling.

**Comment by StrongerXi** [comment]
Plumb a `overridden_sources: Dict[Source, Source]` from `handle_aliases_for_stolen_lists` to `PyCodegen` so that we don't update `VariableTracker.source` in place, while still preserving what `handle_aliases_for_stolen_lists` needed (i.e., modifying codegen for certain `VariableTracker`).

**Comment by StrongerXi** [comment]
Account for the following in `PyCodegen` which bypassed `__call__`:
```c++
    self.call_reconstruct(value.source)
```

**Comment by StrongerXi** [comment]
I did some investigation on main 2b7c7a20b9c:

The `handle_aliases_for_stolen_lists` addressed above is the only exception I could find which breaks the following 2 observations:

1. For all instances of `git grep "\.source ="`, none of them is "mutating" a source field; instead they are all initializing, i.e., setting it from `None` to some `Source` (the initialization might happen outside `__init__`).
2. All `MutableLocal` instances, if given source field, are constructed with the source from their associated `VariableTracker`, meaning they have the same source instance.

File: torch/_dynamo/codegen.py [hunk 2]
```diff
@@ -51,6 +51,7 @@ def __init__(
         root: Optional[torch.nn.Module] = None,
         graph_output_var: Optional[str] = None,
         tempvars=None,
+        overridden_sources=None,
```
Comment (jansel) [inline torch/_dynamo/codegen.py:54]: Add type annotations

File: torch/_dynamo/codegen.py [hunk 3]
```diff
@@ -130,27 +133,24 @@ def __call__(self, value, allow_cache=True):
...
             # mutable_side_effects_from_source=True is used to codegen a reference
             from .side_effects import MutableSideEffects
 
             if isinstance(value.mutable_local, MutableSideEffects):
-                self(value.mutable_local.source)
+                self(value.source)
                 return
 
         if allow_cache:
-            if value.mutable_local and value.mutable_local in self.tempvars:
```
Comment (jansel) [inline torch/_dynamo/codegen.py:143]: This one feels like a separate change from this PR.  Though perhaps it is ok because VariableTrackers are now mutable.

File: torch/_dynamo/codegen.py [hunk 4]
```diff
@@ -116,7 +118,8 @@ def add_push_null(self, gen_fn, call_function_ex=False):
     def __call__(self, value, allow_cache=True):
         """Generate code such that top-of-stack (TOS) is set to value"""
         if isinstance(value, Source):
-            self.call_reconstruct(value)
+            source = self.overridden_sources.get(value, value)
```
Comment (jansel) [inline torch/_dynamo/codegen.py:121]: Need some comments explaining this.

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/codegen.py (hunk 3 above)
Comment (StrongerXi) [reply torch/_dynamo/codegen.py:143]: Yeah this is me trying to remove indexing on `MutableLocal`.

File: torch/_dynamo/codegen.py (hunk 2 above)
Comment (StrongerXi) [reply torch/_dynamo/codegen.py:54]: Updated.

**Comment by StrongerXi** [comment]
Add type annotation and comments to explain `PyCodegen.overridden_sources` and its uses.

File: torch/_dynamo/side_effects.py [hunk 5]
```diff
@@ -36,9 +36,8 @@ class MutableSideEffects(MutableLocalBase):
     the graph runs.
     """
 
-    def __init__(self, source: Source, is_modified: bool = False):
+    def __init__(self, is_modified: bool = False):
```
Comment (anijain2305) [inline torch/_dynamo/side_effects.py:39]: Not relevant to this line and can be done in another PR. Can we move `MutableLocal` and `MutableLocalBase` in this file? Good to consolidate them here, and then add a comment what different types of subclasses mean with some examples.

Comment (anijain2305) [reply torch/_dynamo/side_effects.py:39]: Hmm, moving `MutableLocal` could be messy given so many other files import it.

File: torch/_dynamo/side_effects.py [hunk 6]
```diff
@@ -435,7 +433,7 @@ def codegen_save_tempvars(self, cg: PyCodegen):
                 )
                 cg.extend_output(create_call_function(0, False))
                 cg.add_cache(var)
-                var.mutable_local.source = LocalSource(cg.tempvars[var])  # type: ignore[attr-defined]
+                var.source = LocalSource(cg.tempvars[var])  # type: ignore[attr-defined]
```
Comment (anijain2305) [inline torch/_dynamo/side_effects.py:436]: Trying to understand. Overriding source is OK here because we have already generated guards?

**Review by anijain2305 — APPROVED** [review]
Thanks! Nice work!

**Review by anijain2305 — APPROVED** [review]
Thanks! Nice work!

File: torch/_dynamo/side_effects.py (hunk 5 above)
Comment (StrongerXi) [reply torch/_dynamo/side_effects.py:39]: Yeah I also thought it's a tad odd to have these mutability marker subclasses in different files, as it hinders reasoning about the cases. I'll see what I can do.

File: torch/_dynamo/side_effects.py (hunk 6 above)
Comment (StrongerXi) [reply torch/_dynamo/side_effects.py:436]: Ah so here since we are dealing with temporary variables for which Dynamo supports mutation, e.g.,
```python
def f(x, y):
  tmp_obj = MyObject()
  tmp_obj.foo = x + y
  ...
```
so we don't need guards for `tmp_obj`.

With that I see `Source` being used for 2 purposes in Dynamo:
1. Help generate guards (this is probably a must-have)
2. Help with codegen (this seems to be an impl decision; we could tweak things by, e.g., storing this information as a `Dict[Source, str] source_to_tempvar_name`...)

The use case here is (2). I'll send a subsequent PR to beef up all these documentations, and maybe refactor the impl if that can makes things more clear.

**Comment by StrongerXi** [comment]
Rebase over #135714.

**Review by mlazos — APPROVED** [review]
Looks great!

File: torch/_dynamo/side_effects.py [hunk 7]
```diff
@@ -362,35 +361,30 @@ def track_tensor_variables_from_runahead_side_effects(self, other):
                 self.track_object_existing(other_item, other_variable)
 
     def prune_dead_object_new(self, tx):
-        live_new_objects = set()
+        live_new_objects: Set[VariableTracker] = set()
 
         # use this to avoid cycles in mutable_local (though I'm not sure if that
```
Comment (mlazos) [inline torch/_dynamo/side_effects.py:366]: Wrt to this comment, I don't think cycles can happen here anymore because the VariableTracker to MutableLocal relationship is now 1:1. It might be worth experimenting with removing this, and then we won't need del visit perhaps? Although if the same VT is in multiple places, this might cause some redundant work.

Comment (StrongerXi) [reply torch/_dynamo/side_effects.py:366]: Good call, in my follow up patch I'll at least update the intention here.
